PhyloPyPruner
-------------

PhyloPyPruner is a Python package for refining the output of a graph-based
orthology inference approach such as
[OrthoMCL](https://www.ncbi.nlm.nih.gov/pubmed/12952885),
[OrthoFinder](https://www.ncbi.nlm.nih.gov/pubmed/26243257) or
[HaMStR](https://www.ncbi.nlm.nih.gov/pubmed/19586527). Similar to other
tree-based orthology inference methods (e.g.,
[PhyloTreePruner](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3825643/),
[UPhO](https://academic.oup.com/mbe/article/33/8/2117/2578877),
[Agalma](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3840672/) and
[Phylogenomic Dataset
Reconstruction](https://www.ncbi.nlm.nih.gov/pubmed/25158799), PhyloPyPruner
uses phylogenetic trees in order to obtain 1-to-1 orthologous genes (only a
single copy of that gene is present for each species within the analysis). In
addition to filters and algorithms seen in pre-existing tools, this package
provides new methods for differentiating [contamination-like
sequences](https://gitlab.com/fethalen/phylopypruner/wikis/About-PhyloPyPruner#contamination-like-issues-)
from paralogs.

PhyloPyPruner is under active development and I would appreciate it if you try
this software on your own data and [leave
feedback](mailto:felix.thalen@uni-goettingen.de).

## [Documentation](https://gitlab.com/fethalen/phylopypruner/wikis)

* [About PhyloPyPruner](https://gitlab.com/fethalen/phylopypruner/wikis/about-phylopypruner)
* Tutorial
* [Installation](https://gitlab.com/fethalen/phylopypruner/wikis/installation)
* [Input data](https://gitlab.com/fethalen/phylopypruner/wikis/input-data)
* [Output files](https://gitlab.com/fethalen/phylopypruner/wikis/output-files)
* [Methods](https://gitlab.com/fethalen/phylopypruner/wikis/methods)
* [Options](https://gitlab.com/fethalen/phylopypruner/wikis/options)

## Cite

Manuscript is currently in preparation.

Â© [Kocot Lab](https://www.kocotlab.com/) 2018